Xtensa] Implement windowed call ABI.

andreisfr opened the pull request:
Implement base windowed call ABI. By defaullt use
rotation window by 8 registers.

andreisfr committed:
[Xtensa] Implement windowed call ABI.

Implement base windowed call ABI. By defaullt use
rotation window by 8 registers.

arsenm commented on llvm/lib/Target/Xtensa/XtensaCallingConv.td:
```diff
@@ -9,16 +9,30 @@
+// Callee-saved register lists for rotation window by 8 registers
+def CSRW8_Xtensa : CalleeSavedRegs<(add)> {
+  let OtherPreserved = (add A0, SP, A2, A3, A4, A5, A6, A7);
+}
```
End of file whitespace error

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -23,9 +23,15 @@
 
 using namespace llvm;
 
+/* minimum frame = reg save area (4 words) plus static chain (1 word)
+   and the total number of words must be a multiple of 128 bits.  */
+/* Width of a word, in units (bytes).  */
```
C++ style comments

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -51,6 +58,83 @@ void XtensaFrameLowering::emitPrologue(MachineFunction &MF,
+          .addImm(StackSize);
+    } else {
+      /* Use a8 as a temporary since a0-a7 may be live.  */
+      unsigned TmpReg = Xtensa::A8;
```
MCRegister

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -51,6 +58,83 @@ void XtensaFrameLowering::emitPrologue(MachineFunction &MF,
+      unsigned TmpReg = Xtensa::A8;
+
+      const XtensaInstrInfo &TII = *static_cast<const XtensaInstrInfo *>(
+          MBB.getParent()->getSubtarget().getInstrInfo());
```
Use the STI member you added instead of finding it from the parent function

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -51,6 +58,83 @@ void XtensaFrameLowering::emitPrologue(MachineFunction &MF,
+    // diff_to_128_aligned_address = (128 - (SP & 127))
+    // new_offset = SP + diff_to_128_aligned_address
+    // This is safe to do because we increased the stack size by MaxAlignment.
+    unsigned Reg, RegMisAlign;
```
Use MCRegister type

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -179,10 +264,22 @@ void XtensaFrameLowering::emitEpilogue(MachineFunction &MF,
     }
-
-    BuildMI(MBB, I, DL, TII.get(Xtensa::OR), SP).addReg(FP).addReg(FP);
+    if (STI.isWinABI()) {
+      // In most architectures, we need to explicitly restore the stack pointer
+      // before returning.
+      //
+      // For Xtensa Windowed Register option, it is not needed to explicitly
+      // restore the stack pointer. Reason being is that on function return,
+      // the window of the caller (including the old stack pointer) gets
+      // restored anyways.
+    } else {
+      BuildMI(MBB, I, DL, TII.get(Xtensa::OR), SP).addReg(FP).addReg(FP);
+    }
```
Invert condition and avoid comment only block

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -339,7 +348,18 @@ SDValue XtensaTargetLowering::LowerFormalArguments(
       // physical registers into virtual ones
-      Register Reg = MF.addLiveIn(VA.getLocReg(), &Xtensa::ARRegClass);
+      Register Reg = 0;
+      unsigned FrameReg = Subtarget.getRegisterInfo()->getFrameRegister(MF);
```
MCegister

andreisfr replied:
Fixed.

arsenm commented on llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp:
```diff
@@ -34,13 +34,19 @@ XtensaRegisterInfo::XtensaRegisterInfo(const XtensaSubtarget &STI)
-  return CSR_Xtensa_SaveList;
+  if (Subtarget.isWinABI())
+    return CSRW8_Xtensa_SaveList;
+  else
```
No else after return. Return ternary operator?

arsenm commented on llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp:
```diff
@@ -34,13 +34,19 @@ XtensaRegisterInfo::XtensaRegisterInfo(const XtensaSubtarget &STI)
 XtensaRegisterInfo::getCallPreservedMask(const MachineFunction &MF,
                                          CallingConv::ID) const {
-  return CSR_Xtensa_RegMask;
+  if (Subtarget.isWinABI())
```
Ditto

andreisfr replied:
Fixed.

arsenm commented on llvm/test/CodeGen/Xtensa/calling-conv-windowed.ll:
```diff
@@ -0,0 +1,103 @@
+; XTENSA-NEXT:    l32i a9, a1, 32
+; XTENSA-NEXT:    add a2, a8, a9
+; XTENSA-NEXT:    retw
+  %8 = load i32, ptr %6, align 4
```
Use named values in tests

andreisfr replied:
Corrected

arsenm commented on llvm/lib/Target/Xtensa/XtensaSubtarget.h:
```diff
@@ -70,6 +70,8 @@ class XtensaSubtarget : public XtensaGenSubtargetInfo {
 
   bool hasBoolean() const { return HasBoolean; }
 
+  bool isWinABI() const { return hasWindowed(); }
```
Maybe should expand this to be windowed to avoid casual observers confusing this with windows

andreisfr replied:
Thank you very much for your comments. I changed  "isWInABI" to "isWindowedABI".

andreisfr committed:
[Xtensa] Fix windowed call ABI.

Fixed windowed call ABI test. Code formatting and other minor fixes.

andreisfr committed:
[Xtensa] Minor code formatting.

andreisfr committed:
[Xtensa] Fix XtensaFrameLowering implementation.

andreisfr commented on llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp:
```diff
@@ -51,76 +58,155 @@ void XtensaFrameLowering::emitPrologue(MachineFunction &MF,
+      MCRegister TmpReg = Xtensa::A8;
+
+      const XtensaInstrInfo &TII = *static_cast<const XtensaInstrInfo *>(
+          MBB.getParent()->getSubtarget().getInstrInfo());
```
Thank you. Fixed.

andreisfr committed:
[Xtensa] Implement Windowed ABI for STACKRESTORE.

andreisfr commented:
@arsenm , I missed Windowed ABI implementation for STACKRESTORE operation, sorry. Could you PTAL to the last commit?

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -895,8 +895,14 @@ SDValue XtensaTargetLowering::LowerSTACKSAVE(SDValue Op,
+    SDValue NewSP =
+        DAG.getNode(XtensaISD::MOVSP, SDLoc(Op), MVT::i32, Op.getOperand(1));
+    return DAG.getCopyToReg(Op.getOperand(0), SDLoc(Op), Xtensa::SP, NewSP);
+  } else {
```
No else after return

andreisfr replied:
Fixed

arsenm reviewed (commented):
Missing test change

andreisfr committed:
[Xtensa] Add test for STACKRESTORE. Fix SP update.

Implemented MOVSP_P pseudo operation for SP register update in
STACKRESTORE and DYNAMIC_STACKALLOC lowerinng.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -1385,6 +1389,16 @@ MachineBasicBlock *XtensaTargetLowering::EmitInstrWithCustomInserter(
     return MBB;
   }
+  case Xtensa::MOVSP_P: {
+    MachineOperand SP = MI.getOperand(0);
+    MachineOperand NewSP = MI.getOperand(1);
```
```suggestion
    MachineOperand SP = &MI.getOperand(0);
    MachineOperand NewSP = &MI.getOperand(1);
```

andreisfr replied:
Fixed. I hope that I correctly understand suggestion.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -944,9 +946,11 @@ SDValue XtensaTargetLowering::LowerDYNAMIC_STACKALLOC(SDValue Op,
   SDValue NewSP = DAG.getNode(ISD::SUB, DL, VT, SP, SizeRoundUp); // Value
   if (Subtarget.isWindowedABI()) {
-    NewSP = DAG.getNode(XtensaISD::MOVSP, DL, MVT::i32, NewSP);
+    Chain = DAG.getNode(XtensaISD::MOVSP, SDLoc(Op), MVT::Other, SP.getValue(1),
```
Should this be using the call frame pseudos like other targets? As is I think this will miss setting MachineFrameInfo's AdjustsStack

andreisfr replied:
@arsenm , thank you very much for advice. May I ask to explain a bit more about AdjustStack and call frame pseudos to help me to understand? I investigated several backends and it seems these backends don't use AdjustStack. Maybe I missed something? For example AArch64 and Lanai backends implements lowering function for DYNAMIC_STACKALLOC using specific target dependent pseudo instructions with custom inserting. And proposed solution in this PR for Xtensa is very similar to what is done in AArch64/Lanai and some others backends. Should I remove somehow pseudo instruction with custom inserting or  something else? Thank you.

arsenm commented on llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:
```diff
@@ -1385,6 +1389,16 @@ MachineBasicBlock *XtensaTargetLowering::EmitInstrWithCustomInserter(
+    MachineOperand SP = MI.getOperand(0);
+    MachineOperand NewSP = MI.getOperand(1);
+
+    BuildMI(*MBB, MI, DL, TII.get(Xtensa::MOVSP), SP.getReg())
```
Could this just have used the call frame pseudos to start

andreisfr replied:
I changed MOVSP pseudo instruction to simpler form.

andreisfr committed:
[Xtensa] Fix MOVSP pseudo operation.

andreisfr committed:
[Xtensa] Minor fixes.

andreisfr commented:
@arsenm, sorry to bother you, maybe we can merge this PR in its current state? And improve this code in the future if needed?
